[Profiler] Update README

sraikund16 opened the pull request:
Summary: Updated README with code structure and explanation of core features within profiler

Test Plan:
N/A

Rollback Plan:

Differential Revision: D79604189

sanrise reviewed (approved):
LGTM. Pretty informative even for someone who works on the profiler.

sraikund16 commented on torch/csrc/profiler/README.md:
```diff
@@ -43,14 +76,39 @@ The profiler records two pieces of information from the autograd engine:
+When a trace begins, the autograd/profiler backend calls into `profiler_kineto.cpp` to prepare, start, or stop collection. At the start of tracing, the `onFunctionEnter` and `onFunctionExit` callbacks defined in `profiler_kineto.cpp` are registered.
+
+Callback registration can be either global or local, depending on the `ExperimentalConfig` used:
+- **Global:** The callback is registered to all threads throughout execution.
```
I was thinking of adding this to a tutorial instead so I can add some visuals to the difference?

sraikund16 committed:
[Profiler] Update README (#159816)

Summary:

Updated README with code structure and explanation of core features within profiler

Test Plan:
N/A

Rollback Plan:

Reviewed By: aaronenyeshi

Differential Revision: D79604189
